src/text_shaper/directwrite_locator.cpp

```diff
@@ -0,0 +1,227 @@
+            case DWRITE_FONT_WEIGHT_BLACK: return font_weight::black;
+            case DWRITE_FONT_WEIGHT_EXTRA_BLACK: return font_weight::extra_black;
+            default: // TODO: the others
+                break;
+            }
+            return font_weight::normal; // TODO: rename normal to regular
+        }
+
+        font_slant dwFontSlant(int _style)
+        {
+            switch (_style)
+            {
+            case DWRITE_FONT_STYLE_NORMAL: return font_slant::normal;
+            case DWRITE_FONT_STYLE_ITALIC: return font_slant::italic;
+            case DWRITE_FONT_STYLE_OBLIQUE: return font_slant::oblique;
+            }
+            return font_slant::normal;
+        }
+
+        std::wstring get_font_path(IDWriteFontFace* fontFace)
+        {
+            ComPtr<IDWriteFontFile> fontFile;
+            UINT32 numberOfFiles = 0;
+            fontFace->GetFiles(&numberOfFiles, nullptr);
+            fontFace->GetFiles(&numberOfFiles, &fontFile);
+
+            ComPtr<IDWriteFontFileLoader> loader;
+            fontFile->GetLoader(&loader);
+            const void* key;
+            UINT32 keySize;
+            fontFile->GetReferenceKey(&key, &keySize);
+            ComPtr<IDWriteLocalFontFileLoader> localLoader;
+            loader.As(&localLoader);
+
+            UINT32 pathLen;
+            localLoader->GetFilePathLengthFromKey(key, keySize, &pathLen);
+            std::wstring path;
+            path.resize(pathLen);
+            localLoader->GetFilePathFromKey(key, keySize, path.data(), pathLen + 1);
+
+            return path;
+        }
+    } // }}}
+
+    struct directwrite_locator::Private
+    {
+        ComPtr<IDWriteFactory7> factory;
+        ComPtr<IDWriteFontCollection> systemFontCollection;
+        ComPtr<IDWriteFontFallback> systemFontFallback;
+
+        std::wstring userLocale;
+
+        Private()
+        {
+            auto hr = DWriteCreateFactory(DWRITE_FACTORY_TYPE_SHARED,
+                __uuidof(IDWriteFactory7),
+                reinterpret_cast<IUnknown**>(factory.GetAddressOf()));
+
+            wchar_t locale[LOCALE_NAME_MAX_LENGTH];
+            GetUserDefaultLocaleName(locale, sizeof(locale));
+            userLocale = locale;
+
+            factory->GetSystemFontCollection(&systemFontCollection);
+            factory->GetSystemFontFallback(&systemFontFallback);
+        }
+    };
+
+    directwrite_locator::directwrite_locator() :
+        d{ new Private(), [](Private* p) { delete p; } }
+    {
+    }
+
+    directwrite_locator::~directwrite_locator()
+    {
+    }
+
+    font_source_list directwrite_locator::locate(font_description const& _fd)
+    {
+        font_source_list output;
+
+        // TODO: use libunicode for that (TODO: create wchar_t/char16_t converters in libunicode)
+        std::wstring_convert<std::codecvt_utf8_utf16<wchar_t>> wStringConverter;
+        std::wstring familyName = wStringConverter.from_bytes(_fd.familyName);
```

Comment:
The conversion between UTF8 <-> UTF16 is just everywhere.....

Reply:
The layer above, that is using the <text_shaper> API is caching, so only uncache requests come here.

While that is good, I think we should try to get rid if unnecessary conversions, I agree. But that can be easily done in a separate PR that can be used for performance optimizations (such as this).

Also I have created a ticket in libunicode to fix UTF16/32 conversion. (the C++ API is actually deprecated in the C++ standard, I also do not know why, but yeah).